metricbeat/helper/prometheus/prometheus.go

```diff
@@ -76,110 +86,224 @@ func (p *prometheus) GetFamilies() ([]*dto.MetricFamily, error) {
 	}
 
 	families := []*dto.MetricFamily{}
-	for {
-		mf := &dto.MetricFamily{}
-		err = decoder.Decode(mf)
-		if err != nil {
-			if err == io.EOF {
-				break
+	var wg sync.WaitGroup
+	var decoderMutex = &sync.Mutex{}
+	var familiesMutex = &sync.Mutex{}
+	var done = false
+	var count = 0
+	for !done && count < maxGetFamiliesSubFuns {
+		wg.Add(1)
+		count++
+		go func(id int) {
+			for !done {
+				mf := &dto.MetricFamily{}
+				decoderMutex.Lock()
+				err = decoder.Decode(mf)
```

Comment:
probably cuter to make your own decoder that encapsulates expfmt.decoder and bundles it with your mutexes. then these calls will just be  gernotsDecoder.Decode(mf) and the mutex locking/unlocking will be hidden inside the API. then users wont make the potential mistake of dealing with these mutexes.

Reply:
Not sure of the benefit here since this is localized to the prometheus package. 
The invocation of the Decoder is private to the prometheus package.

Reply:
up to you.  there's no explicit benefit - just more safeguarding against mistakes by people modifying this code in the future.